src/Models/HydrostaticFreeSurfaceModels/implicit_free_surface_solver.jl

```diff
@@ -0,0 +1,77 @@
+using Oceananigans.Solvers
+using Oceananigans.Operators
+
+struct ImplicitFreeSurfaceSolver{S}
+    solver   :: S
+end
+
+function ImplicitFreeSurfaceSolver(arch, template_field, 
+                                   vertically_integrated_lateral_face_areas;
+                                   Amatrix_operator=nothing, 
+                                   maxit=nothing, 
+                                   tol=nothing)
+       
+    ## This should probably end up in some operator and grid generic auxilliaries
+    ## NEED TO SORT OUT HALO FOR PERIODIC CASE
+    Ax_baro = vertically_integrated_lateral_face_areas.Ax
+    Ay_baro = vertically_integrated_lateral_face_areas.Ay
+    @inline Ax_∂xᶠᵃᵃ_baro(i, j, k, grid, c) = Ax_baro[i, j, 1] * ∂xᶠᵃᵃ(i, j, 1, grid, c)
+    @inline Ay_∂yᵃᶠᵃ_baro(i, j, k, grid, c) = Ay_baro[i, j, 1] * ∂yᵃᶠᵃ(i, j, 1, grid, c)
+    @inline function ∇²_baro(i, j, k, grid, c)
+       return  δxᶜᵃᵃ(i, j, 1, grid, Ax_∂xᶠᵃᵃ_baro, c) +
+               δyᵃᶜᵃ(i, j, 1, grid, Ay_∂yᵃᶠᵃ_baro, c)
+    end
+
+    @kernel function implicit_η!(grid, f, implicit_η_f)
+        i, j = @index(Global, NTuple)
+        ### Not sure what to call this
+        ### it is for left hand side operator in
+        ### ( ∇ʰ⋅H∇ʰ - 1/gΔt² ) ηⁿ⁺¹ = 1/(gΔt) ∇ʰH U̅ˢᵗᵃʳ - 1/(gΔt²) ηⁿ
+        ### written in a discrete finite volume form in which the equation
+        ### is arranged to ensure a symmtric form
+        ### e.g.
+        ### 
+        ### δⁱÂʷ∂ˣηⁿ⁺¹ + δʲÂˢ∂ʸηⁿ⁺¹ - 1/gΔt² Aᶻηⁿ⁺¹ =
+        ###  1/(gΔt)(δⁱÂʷu̅ˢᵗᵃʳ + δʲÂˢv̅ˢᵗᵃʳ) - 1/gΔt² Aᶻηⁿ
+        ###
+        ### where  ̂ indicates a vertical integral, and
+        ###        ̅ indicates a vertical average
+        ###
+        g  = Main.model.free_surface.gravitational_acceleration ### AGHHHH - need to sort this out later.....
```

Comment:
😂 

Perhaps `g` should be passed directly into the kernel function as an argument?

Reply:
(I didn't know you could access children of `Main` this way...)

Reply:
Good idea :-)

I couldn't decide whether you and @ali-ramadhan would like a "context" argument for the generic A operator (which is used through a generic interface here https://github.com/CliMA/Oceananigans.jl/blob/a211d9d0f0ddf3f66ca9fe4319f47aac1e9e6164/src/Solvers/preconditioned_conjugate_gradient_solver.jl#L173 ) or whether you get all excited about types and structs and decide "p" should be a type. 

I thought the easiest way to figure that out would be to use Main. for now!

Reply:
> (I didn't know you could access children of `Main` this way...)

I don't think it is recommended practice - but pretty much everything in Julia can be reached from anywhere, when needed!

Reply:
That's very scary.